### ISSUE TITLE
[META] Windows/MacOS support mega thread

### ISSUE THREAD

**Issue by ehfd** [selkies-project/selkies#57]
**NOTE: We will not accept a PR regarding this issue right now, due to the volume of code it modifies, unless there is a dedicated and determined maintainer to continue maintaining these components. There are free (but not necessarily open source) alternatives in these OSes, including Reemo and Parsec, for now.**

**NOTE: This needs a maintainer dedicated to each or both OS. Current maintainers are focused on Linux usage.**

---

Can be AI-driven: Yes, as long as all cases are tested carefully in real-life by people as well as automated tests. However, **expert guidance is required due to working with OS intrinsics and media pipelines**.
Expected to be workable with: Fable 5, GPT-5.6 Sol, or equivalent (max thinking, requires expert guidance, especially with Opus 5, Kimi-K3, or under).

Update: https://crates.io/crates/screencapturekit (and avf as fallback) should be used for macOS. For Windows, we aim to support the latest versions of Windows 11 (d3d12), with another API that supports later Windows 10 versions (d3d11/mf; the two of these are actually hardware compatibility as far as I'm aware) as a fallback. The best API for this case, which uses zero-copy pipelines, should be used. For audio, macOS either uses the API in screencapturekit or CoreAudio; Windows uses the Windows Audio Session API available with Windows 10. If there is any other API that is better, need to use it.

---

Windows support is in progress with the contribution from the community.

Note that the below are based on GStreamer conventions and we now use pixelflux and pcmflux.

For Windows, the pipeline should add `d3d11screencapturesrc` (screen capture), `d3d11colorconvert` (color convert), and `wasapi` or `wasapi2` (audio capture).

In addition, it would be worth adding the `mfh264enc` and `mfvp9enc` encoder pipelines for Windows native encoding capabilities, as well as zero-copy `nvd3d11h264enc`, `nvd3d11h265enc`, and `nvd3d11av1enc` for NVIDIA, which also may be used for Windows.

For MacOS, `pyobjc-framework-applicationservices` and `pyobjc-framework-quartz` are possibly frequently used interfaces for input.

ScreenCaptureKit is preferred with `avfvideosrc capture-screen=true` as fallback(?), `glcolorconvert` (color convert, not sure), and `osxaudiosrc` (audio capture) are probably the way to go for media capture.

**Comment by matbeedotcom** [comment]
// Removal of patches to the GStreamer build process in https://github.com/selkies-project/selkies-gstreamer/pull/51 means that the binary build available from the project for Windows will work without additional alterations.

Does this mean I can currently run a compile of the project while removing the patches to make a windows build?

**Answer from maintainer: Yes, but the capture interface and input interface are not written.**

**Comment by ehfd** [comment]
Does not work out of the box right now. BUT, we have contributors @callTx and @Xosrov doing this and probably have it about 80% working.

**Comment by ehfd** [comment]
https://discord.gg/wDNGDeSW5F

We have a channel specifically for Windows development.

**Comment by boxerab** [comment]
Hi folks, what's the status of Windows development ?

**Comment by ehfd** [comment]
@callTx

**Comment by boxerab** [comment]
@ehfd would you be able to point me to a branch with the windows port ? I looked at the forks but couldn't find anything windows specific.

**Comment by ehfd** [comment]
@boxerab https://github.com/selkies-project/selkies-gstreamer-msys2
TBA

**Comment by boxerab** [comment]
@ehfd still interested in seeing code for the windows port.  Any updates on when that might be available ?
Thanks!

**Comment by boxerab** [comment]
hmmmmmm - I would very much like to see any selkies code ported to windows.

**Comment by ehfd** [comment]
@callTx

**Comment by callTx** [comment]
Hello @ehfd and all. Really sorry for the Windows port has not been
released. I was facing on a plenty of gstreamer connection bug on my local
repo and after many fail attempts I decide refactore and revise all code,
most of them on gstreamer webrtc core.

Em qui., 27 de abr. de 2023 às 23:13, Seungmin Kim ***@***.***>
escreveu:

**Comment by callTx** [comment]
Don't worry guys I'll not give up this project.

Em sex., 28 de abr. de 2023 às 08:43, callTX ***@***.***>
escreveu:

**Comment by boxerab** [comment]
@callTx thanks, looking forward to this. Even though not complete, why don't you share the code on a branch? I'm very keen to take a look. Also, I may be able to help with port.

**Comment by ne0YT** [comment]
@callTx 
hey there,
any updates?

**Comment by ehfd** [comment]
The code has become too Linux-centric. Whoever wants to do this must clean up the code to distinguish elements between Windows and Linux.

**Comment by wouldntyouknow** [comment]
Were there any advancements on this, anyone knows?

**Comment by ehfd** [comment]
Initial Windows support:
https://github.com/callTx/selkies-gstreamer/tree/feature/windows-nvenc-port

**Comment by ehfd** [comment]
Instructions on what I will accept from a PR for Windows support:

1. Start from here; this is about 70% done: https://github.com/callTx/selkies-gstreamer/tree/feature/windows-nvenc-port

2. Implement D3D11 screen capture and both a zero-copy hardware NVIDIA pipeline (if possible and have the hardware; what I mean here is that the implementation which downloads from D3D11 and then encodes is sub-optimal), and a general x264 pipeline as well, in PixelFlux.

3. Implement Windows sound capture on PCMFlux.

Only when the three constraints are fulfilled, and there are no reintroduction of GStreamer, can we accept a Windows support PR.

**Comment by callTx** [comment]
1. I can move forward with implementing D3D11/DXGI screen capture with a zero-copy NVENC path and a CPU-based x264 fallback in PixelFlux. 
2. Since PCMFlux is currently Linux-exclusive, what do you think in plan to develop a Windows-native module using WASAPI to replicate the audio capture functionality?

**Comment by ne0YT** [comment]
Is there any way this can smh be used to make windows docker containers work for GUI apps?

**Comment by callTx** [comment]
@ne0YT AFIK  Windows container images (like Server Core or Nano Server) lack the necessary Desktop Window Manager (DWM) and GUI APIs required to draw and display window. Can you confirm me @ehfd ?

**Comment by ne0YT** [comment]
@callTx maybe possible to inject what's needed?

**Comment by callTx** [comment]
I think so @ne0YT

**Comment by ehfd** [comment]
Not sure from my side.

**Comment by kingdudely** [comment]
Why not use ScreenCaptureKit instead of AVFoundation for MacOS?

**Comment by ehfd** [comment]
@kingdudely Good idea. Since Selkies now uses Pixelflux and PCMflux over GStreamer, there are no longer GStreamer limitations.
